# Cleaning the Restaurants Dataset
The goal of this notebook is to clean and prepare the raw restaurant dataset for analysis. This includes handling missing values, standardizing data formats, and engineering relevant features related to restaurant characteristics such as ratings, opening hours, and available services.

The cleaned dataset produced here will be used for analyzing the key factors that contribute to the success of restaurants in Edmonton.

## 1. Import Libraries and Data

In this step, the dataset was loaded and a preliminary inspection was conducted to understand its structure and quality. This includes examining the dataset shape, data types, missing values, duplicated records, and basic descriptive statistics. The restaurant data was acquired from Place API (details refer to [02_get_restaurant_data.ipynb](02_get_restaurant_data.ipynb)).

The goal of this step is to identify potential data quality issues and determine the appropriate data cleaning strategy.

In [1]:
import pandas as pd
from datetime import time

pd.set_option('display.max_columns', None)
pd.set_option('future.no_silent_downcasting', True)

In [2]:
df = pd.read_json('data/raw_yeg_restaurants.json')
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,displayName,primaryTypeDisplayName,takeout,delivery,dineIn,curbsidePickup,reservable,servesBreakfast,servesLunch,servesDinner,servesBeer,servesWine,servesBrunch,primaryType,outdoorSeating,liveMusic,menuForChildren,servesCocktails,servesDessert,servesCoffee,goodForChildren,allowsDogs,restroom,goodForGroups,goodForWatchingSports,paymentOptions,parkingOptions,accessibilityOptions,priceRange,servesVegetarianFood,editorialSummary
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_INEXPENSIVE,749.0,{'text': 'Ecity Flame (Best Offer on Website)'...,"{'text': 'Restaurant', 'languageCode': 'en-US'}",1.0,1.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,restaurant,0.0,0.0,1.0,1.0,1.0,1.0,1.0,0.0,1.0,1.0,1.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'paidParkingLot': Fal...","{'wheelchairAccessibleParking': True, 'wheelch...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,None
1,ChIJQ2rvH4cYoFMRsV6eIIgjmRs,"[gas_station, convenience_store, fast_food_res...","7535 75 Street NW, Edmonton, AB T6C 4M3, Canada",3.4,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_MODERATE,95.0,"{'text': 'Shell', 'languageCode': 'en'}","{'text': 'Gas Station', 'languageCode': 'en-US'}",1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,gas_station,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",{'wheelchairAccessibleRestroom': True},None,NaN,None
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_MODERATE,884.0,"{'text': 'JT's Bar & Grill', 'languageCode': '...","{'text': 'Bar & Grill', 'languageCode': 'en-US'}",1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,bar_and_grill,1.0,1.0,0.0,1.0,1.0,1.0,0.0,0.0,1.0,1.0,1.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'wheelchairAccessibleParking': True, 'wheelch...","{'startPrice': {'currencyCode': 'CAD', 'units'...",1.0,None
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_INEXPENSIVE,84.0,"{'text': 'Little Caesars Pizza', 'languageCode...","{'text': 'Pizza Restaurant', 'languageCode': '...",1.0,1.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,pizza_restaurant,0.0,0.0,0.0,0.0,NaN,0.0,1.0,0.0,NaN,NaN,0.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'wheelchairAccessibleParking': True, 'wheelch...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,{'text': 'Carry-out chain featuring chicken wi...
4,ChIJiXfDGDQhoFMRehR_Gj75DHs,"[convenience_store, hotel, food_store, lodging...","16806 118 Ave NW, Edmonton, AB T5V 1M8, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",None,15.0,"{'text': 'West Edmonton Truckland Ltd', 'langu...","{'text': 'Convenience Store', 'languageCode': ...",1.0,1.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,convenience_store,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'wheelchairAccessibleParking': True, 'wheelch...",None,NaN,None


In [3]:
print(f"Number of duplicated records: {df.duplicated('id').sum()}")

Number of duplicated records: 0


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 826 entries, 0 to 825
Data columns (total 38 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      826 non-null    object 
 1   types                   826 non-null    object 
 2   formattedAddress        826 non-null    object 
 3   rating                  821 non-null    float64
 4   regularOpeningHours     800 non-null    object 
 5   priceLevel              673 non-null    object 
 6   userRatingCount         821 non-null    float64
 7   displayName             826 non-null    object 
 8   primaryTypeDisplayName  824 non-null    object 
 9   takeout                 782 non-null    float64
 10  delivery                763 non-null    float64
 11  dineIn                  742 non-null    float64
 12  curbsidePickup          324 non-null    float64
 13  reservable              600 non-null    float64
 14  servesBreakfast         529 non-null    float64

In [5]:
df.describe(include='all')

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,displayName,primaryTypeDisplayName,takeout,delivery,dineIn,curbsidePickup,reservable,servesBreakfast,servesLunch,servesDinner,servesBeer,servesWine,servesBrunch,primaryType,outdoorSeating,liveMusic,menuForChildren,servesCocktails,servesDessert,servesCoffee,goodForChildren,allowsDogs,restroom,goodForGroups,goodForWatchingSports,paymentOptions,parkingOptions,accessibilityOptions,priceRange,servesVegetarianFood,editorialSummary
count,826,826,826,821.000000,800,673,821.000000,826,824,782.000000,763.000000,742.000000,324.000000,600.000000,529.000000,728.000000,716.00000,700.000000,682.000000,398.000000,824,680.000000,643.000000,604.000000,578.000000,633.000000,630.000000,624.000000,712.000000,716.000000,546.000000,593.000000,814,779,805,633,496.00000,454
unique,826,320,807,NaN,505,4,NaN,479,59,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,59,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,13,30,32,10,NaN,177
top,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[restaurant, point_of_interest, food, establis...","6655 178 St NW, Edmonton, AB T5T 4J5, Canada",NaN,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_MODERATE,NaN,"{'text': 'Tim Hortons', 'languageCode': 'en'}","{'text': 'Restaurant', 'languageCode': 'en-US'}",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,restaurant,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'wheelchairAccessibleParking': True, 'wheelch...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,{'text': 'Canadian chain for signature premium...
freq,1,53,3,NaN,70,334,NaN,58,186,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,186,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,745,625,533,372,NaN,54
mean,NaN,NaN,NaN,4.029842,NaN,NaN,836.557856,NaN,NaN,0.997442,0.900393,0.951482,0.651235,0.478333,0.517958,0.980769,0.98324,0.454286,0.369501,0.517588,NaN,0.227941,0.055988,0.703642,0.380623,0.939968,0.787302,0.897436,0.014045,0.960894,0.992674,0.172007,NaN,NaN,NaN,NaN,0.96371,NaN
std,NaN,NaN,NaN,0.487126,NaN,NaN,1087.228815,NaN,NaN,0.050540,0.299671,0.215002,0.477317,0.499947,0.500150,0.137430,0.12846,0.498262,0.483024,0.500320,NaN,0.419813,0.230077,0.457029,0.485961,0.237733,0.409541,0.303632,0.117759,0.193983,0.085356,0.377705,NaN,NaN,NaN,NaN,0.18720,NaN
min,NaN,NaN,NaN,2.200000,NaN,NaN,2.000000,NaN,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,NaN,NaN,NaN,0.00000,NaN
25%,NaN,NaN,NaN,3.800000,NaN,NaN,233.000000,NaN,NaN,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000,1.00000,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,0.000000,NaN,NaN,NaN,NaN,1.00000,NaN
50%,NaN,NaN,NaN,4.100000,NaN,NaN,513.000000,NaN,NaN,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.00000,0.000000,0.000000,1.000000,NaN,0.000000,0.000000,1.000000,0.000000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,0.000000,NaN,NaN,NaN,NaN,1.00000,NaN
75%,NaN,NaN,NaN,4.400000,NaN,NaN,1038.000000,NaN,NaN,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000,1.000000,1.000000,1.000000,NaN,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,0.000000,NaN,NaN,NaN,NaN,1.00000,NaN


In [6]:
df.isna().sum().sort_values(ascending=False)[df.isna().sum()>0]

curbsidePickup            502
servesBrunch              428
editorialSummary          372
servesVegetarianFood      330
servesBreakfast           297
goodForGroups             280
servesCocktails           248
goodForWatchingSports     233
reservable                226
menuForChildren           222
goodForChildren           202
servesCoffee              196
servesDessert             193
priceRange                193
liveMusic                 183
priceLevel                153
outdoorSeating            146
servesWine                144
servesBeer                126
allowsDogs                114
restroom                  110
servesDinner              110
servesLunch                98
dineIn                     84
delivery                   63
parkingOptions             47
takeout                    44
regularOpeningHours        26
accessibilityOptions       21
paymentOptions             12
userRatingCount             5
rating                      5
primaryType                 2
primaryTyp

## 2. Data Cleaning Plan

Based on the initial inspection of the dataset, the following data cleaning and transformation steps were planned to improve data quality and prepare the data for analysis:
- Remove unnecessary and irrelevant columns to reduce noise
- Convert boolean-like columns stored as numeric values into proper boolean type
- Standardize text-based columns to ensure consistency
- Remove records that do not represent restaurants
- Extract relevant fields from dictionary-based columns
- Extract Forward Sortation Area (FSA) codes from formatted address for geographic analysis
- Handle missing values based on column meaning and analytical importance
- Perform a final column cleanup and rearrange columns for better readability

## 3. Remove Unnecessary Columns

This cleaning is to prepare the dataset for analysis that focuses on variables a new restaurant owner can realistically control including pricing, operation hours, location choice, and target customer positioning.

Therefore, columns that were not relevant to the analysis objectives were removed to reduce noise and simplify the dataset. This helps improve readability and focus on variables that contribute meaningful information.

In [7]:
drop_cols = ['primaryTypeDisplayName', 'takeout', 'delivery', 'dineIn', 'curbsidePickup', 'reservable', 'servesBreakfast', 'servesLunch',
             'servesDinner','servesBeer', 'servesWine', 'servesBrunch', 'outdoorSeating', 'liveMusic', 'servesCocktails', 'servesDessert',
             'servesCoffee', 'allowsDogs', 'restroom', 'goodForWatchingSports','accessibilityOptions', 'servesVegetarianFood',
             'editorialSummary']

df.drop(columns=drop_cols, inplace=True)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,displayName,primaryType,menuForChildren,goodForChildren,goodForGroups,paymentOptions,parkingOptions,priceRange
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_INEXPENSIVE,749.0,{'text': 'Ecity Flame (Best Offer on Website)'...,restaurant,1.0,1.0,1.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'paidParkingLot': Fal...","{'startPrice': {'currencyCode': 'CAD', 'units'..."
1,ChIJQ2rvH4cYoFMRsV6eIIgjmRs,"[gas_station, convenience_store, fast_food_res...","7535 75 Street NW, Edmonton, AB T6C 4M3, Canada",3.4,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_MODERATE,95.0,"{'text': 'Shell', 'languageCode': 'en'}",gas_station,NaN,NaN,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",None
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_MODERATE,884.0,"{'text': 'JT's Bar & Grill', 'languageCode': '...",bar_and_grill,0.0,0.0,1.0,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'..."
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",PRICE_LEVEL_INEXPENSIVE,84.0,"{'text': 'Little Caesars Pizza', 'languageCode...",pizza_restaurant,0.0,1.0,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'..."
4,ChIJiXfDGDQhoFMRehR_Gj75DHs,"[convenience_store, hotel, food_store, lodging...","16806 118 Ave NW, Edmonton, AB T5V 1M8, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",None,15.0,"{'text': 'West Edmonton Truckland Ltd', 'langu...",convenience_store,NaN,NaN,NaN,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",None


## 4. Convert Boolean Columns from Float

Several columns were stored as numeric values (e.g. 0.0 and 1.0) but represented boolean information.
These columns were converted to proper boolean type to ensure logical consistency and prevent ambiguity during analysis.

In [8]:
bool_cols = ['menuForChildren', 'goodForGroups', 'goodForChildren']

df[bool_cols] = (
    df[bool_cols]
        .replace({1: True, 0: False})
        .fillna(False)
        .astype(bool)
)

df[bool_cols].info()

<class 'pandas.core.frame.DataFrame'>
Index: 826 entries, 0 to 825
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   menuForChildren  826 non-null    bool 
 1   goodForGroups    826 non-null    bool 
 2   goodForChildren  826 non-null    bool 
dtypes: bool(3)
memory usage: 8.9 KB


## 5. Standardize Text Columns

Text-based columns were standardized by converting all text to lowercase and normalizing inconsistent labels where applicable.
This prevents duplicated categories caused by inconsistent text formatting.

### Price Level Column

In [9]:
df['priceLevel'].unique()

array(['PRICE_LEVEL_INEXPENSIVE', 'PRICE_LEVEL_MODERATE', None,
       'PRICE_LEVEL_EXPENSIVE', 'PRICE_LEVEL_VERY_EXPENSIVE'],
      dtype=object)

In [10]:
df['priceLevel'] = (
    df['priceLevel']
        .str.split('_', n=2)
        .str[-1]
        .str.replace('_',' ')
        .str.lower()
)

df['priceLevel'].unique()

array(['inexpensive', 'moderate', None, 'expensive', 'very expensive'],
      dtype=object)

### Primary Type Column

In [11]:
df['primaryType'].unique()

array(['restaurant', 'gas_station', 'bar_and_grill', 'pizza_restaurant',
       'convenience_store', 'sushi_restaurant', 'meal_delivery',
       'japanese_restaurant', 'meal_takeaway', 'amusement_center',
       'community_center', 'bar', 'sandwich_shop',
       'vietnamese_restaurant', 'steak_house', 'fast_food_restaurant',
       'chinese_restaurant', 'coffee_shop', 'cafe', 'italian_restaurant',
       'breakfast_restaurant', 'mexican_restaurant', 'market',
       'asian_restaurant', 'greek_restaurant',
       'middle_eastern_restaurant', 'korean_restaurant',
       'thai_restaurant', 'turkish_restaurant', 'american_restaurant',
       'bakery', 'indian_restaurant', 'hamburger_restaurant', 'diner',
       'grocery_store', 'sports_club', 'seafood_restaurant',
       'golf_course', 'comedy_club', 'mediterranean_restaurant',
       'juice_shop', 'ice_cream_shop', 'night_club', 'pub',
       'vegan_restaurant', 'bowling_alley', 'casino', 'hotel',
       'tourist_attraction', 'buffet_rest

In [12]:
df['primaryType'] = (
    df['primaryType']
        .str.replace('_',' ')
        .str.lower()
)
df['primaryType'].unique()

array(['restaurant', 'gas station', 'bar and grill', 'pizza restaurant',
       'convenience store', 'sushi restaurant', 'meal delivery',
       'japanese restaurant', 'meal takeaway', 'amusement center',
       'community center', 'bar', 'sandwich shop',
       'vietnamese restaurant', 'steak house', 'fast food restaurant',
       'chinese restaurant', 'coffee shop', 'cafe', 'italian restaurant',
       'breakfast restaurant', 'mexican restaurant', 'market',
       'asian restaurant', 'greek restaurant',
       'middle eastern restaurant', 'korean restaurant',
       'thai restaurant', 'turkish restaurant', 'american restaurant',
       'bakery', 'indian restaurant', 'hamburger restaurant', 'diner',
       'grocery store', 'sports club', 'seafood restaurant',
       'golf course', 'comedy club', 'mediterranean restaurant',
       'juice shop', 'ice cream shop', 'night club', 'pub',
       'vegan restaurant', 'bowling alley', 'casino', 'hotel',
       'tourist attraction', 'buffet rest

## 6. Remove records that are not restaurants

According to the Primary Type column, the dataet contained records that did not represent restaurants, such as gas station and convenience store.
These records were filtered out to ensure the analysis focuses only on relevant business entities.

In [13]:
df[df['primaryType'].isna()]

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,displayName,primaryType,menuForChildren,goodForChildren,goodForGroups,paymentOptions,parkingOptions,priceRange
587,ChIJY_id0_IjoFMRRCg5MEovvMs,"[restaurant, food, point_of_interest, establis...","12535 126 St NW, Edmonton, AB T5L 0X4, Canada",4.7,"{'openNow': False, 'periods': [{'open': {'day'...",None,49.0,"{'text': 'North Star Repair', 'languageCode': ...",None,False,False,False,"{'acceptsCreditCards': True, 'acceptsCashOnly'...",{'freeStreetParking': True},None
607,ChIJ49pJUCMioFMR8UAbkkCPvJg,"[wedding_venue, gym, event_venue, sports_compl...","11160 River Valley Rd NW, Edmonton, AB T5J 2G7...",4.4,"{'openNow': True, 'periods': [{'open': {'day':...",None,420.0,"{'text': 'Royal Glenora Club', 'languageCode':...",None,False,False,False,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",None


For the data without primary type provided, it is hard to determine whether it is a restaurant. Therefore, these 2 records were dropped from the dataset.

In [14]:
df.dropna(subset=['primaryType'], inplace=True, ignore_index=True)

In [15]:
not_restaurants = ['gas station', 'convenience store', 'amusement center','community center', 'market', 'grocery store', 'sports club',
                   'golf course', 'comedy club', 'bowling alley', 'casino', 'hotel', 'tourist attraction', 'food store','shopping mall',
                   'furniture store','store', 'butcher shop', 'performing arts theater']

df = df[df['primaryType'].apply(lambda x: x not in not_restaurants)]
df['primaryType'].unique()

array(['restaurant', 'bar and grill', 'pizza restaurant',
       'sushi restaurant', 'meal delivery', 'japanese restaurant',
       'meal takeaway', 'bar', 'sandwich shop', 'vietnamese restaurant',
       'steak house', 'fast food restaurant', 'chinese restaurant',
       'coffee shop', 'cafe', 'italian restaurant',
       'breakfast restaurant', 'mexican restaurant', 'asian restaurant',
       'greek restaurant', 'middle eastern restaurant',
       'korean restaurant', 'thai restaurant', 'turkish restaurant',
       'american restaurant', 'bakery', 'indian restaurant',
       'hamburger restaurant', 'diner', 'seafood restaurant',
       'mediterranean restaurant', 'juice shop', 'ice cream shop',
       'night club', 'pub', 'vegan restaurant', 'buffet restaurant',
       'french restaurant', 'tea house', 'lebanese restaurant'],
      dtype=object)

## 7. Group Similar Restaurant Primary Types

#### Infer Specific Restaurant Types from Generic "Restaurant" Labels
The majority of restaurants have a primary type labeled as "restaurant", which lacks sufficient specificity for analysis.
To better capture the nature of each establishment, the types column was leveraged to infer more granular restaurant categories.

In [16]:
df['primaryType'].value_counts().sort_values(ascending=False)

primaryType
restaurant                   186
pizza restaurant             101
fast food restaurant          91
sandwich shop                 57
coffee shop                   57
chinese restaurant            36
hamburger restaurant          27
italian restaurant            21
meal delivery                 20
vietnamese restaurant         19
indian restaurant             17
japanese restaurant           15
breakfast restaurant          13
pub                           10
thai restaurant               10
greek restaurant               9
juice shop                     8
bar and grill                  8
bar                            7
mexican restaurant             7
asian restaurant               7
seafood restaurant             6
meal takeaway                  5
sushi restaurant               5
steak house                    5
cafe                           5
middle eastern restaurant      4
korean restaurant              4
turkish restaurant             3
american restaurant            

In [17]:
primary_types = set(df['primaryType'].unique())
primary_types.discard('restaurant')

types = df['types'].explode().str.replace('_', ' ')
types = types[(types.isin(primary_types)) & (df['primaryType'] == 'restaurant')]
types

0      breakfast restaurant
11               night club
11                      bar
26         pizza restaurant
26              steak house
               ...         
770    fast food restaurant
770           meal takeaway
791         thai restaurant
791        asian restaurant
820    fast food restaurant
Name: types, Length: 187, dtype: object

In [18]:
df['possible_types'] = types.groupby(level=0).first()
df.loc[(df['primaryType']=='restaurant') & df['possible_types'], 'primaryType'] = df['possible_types']

df['primaryType'].value_counts()

primaryType
fast food restaurant         110
pizza restaurant             105
restaurant                    95
sandwich shop                 61
coffee shop                   58
chinese restaurant            40
hamburger restaurant          30
meal delivery                 22
italian restaurant            21
vietnamese restaurant         20
breakfast restaurant          19
indian restaurant             18
japanese restaurant           15
bar                           14
bar and grill                 13
pub                           12
thai restaurant               12
night club                    12
asian restaurant              10
greek restaurant               9
juice shop                     8
mexican restaurant             8
steak house                    7
meal takeaway                  7
seafood restaurant             6
vegan restaurant               6
sushi restaurant               5
cafe                           5
korean restaurant              5
middle eastern restaurant      

#### Create a Refined cuisine_group from primaryType
The original primaryType column contains highly granular restaurant categories (e.g. sushi restaurant, pizza restaurant, bar and grill). To reduce fragmentation while preserving meaningful distinctions, similar restaurant types were first grouped into a refined cuisine_group.

This step consolidates closely related business types under a more interpretable category.
For example:
- sushi restaurant and japanese restaurant → japanese restaurant
- pizza restaurant and italian restaurant → italian restaurant
- coffee shop and cafe → beverage shop

Restaurant types that did not belong to any predefined group were retained using their original primaryType to avoid information loss.

In [19]:
american = ['steak house', 'hamburger restaurant', 'diner', 'american restaurant']
italian = ['italian restaurant', 'pizza restaurant']
japanese = ['sushi restaurant', 'japanese restaurant']
middle_eastern = ['middle eastern restaurant', 'lebanese restaurant', 'turkish restaurant']
bar = ['bar','bar and grill', 'night club', 'pub']
beverage = ['coffee shop', 'cafe', 'tea house', 'juice shop']
bakery_dessert = ['bakery', 'ice cream shop']
fast_food = ['fast food restaurant', 'sandwich shop']

df.loc[df['primaryType'].isin(american), 'cuisine_group'] = 'american restaurant'
df.loc[df['primaryType'].isin(italian), 'cuisine_group'] = 'italian restaurant'
df.loc[df['primaryType'].isin(japanese), 'cuisine_group'] = 'japanese restaurant'
df.loc[df['primaryType'].isin(middle_eastern), 'cuisine_group'] = 'middle eastern restaurant'
df.loc[df['primaryType'].isin(bar), 'cuisine_group'] = 'bar'
df.loc[df['primaryType'].isin(beverage), 'cuisine_group'] = 'beverage shop'
df.loc[df['primaryType'].isin(bakery_dessert), 'cuisine_group'] = 'bakery and dessert'
df.loc[df['primaryType'].isin(fast_food), 'cuisine_group'] = 'fast food restaurant'
df['cuisine_group'] = df['cuisine_group'].fillna(df.loc[df['cuisine_group'].isna(), 'primaryType'])
df['cuisine_group'].value_counts()

cuisine_group
fast food restaurant         171
italian restaurant           126
restaurant                    95
beverage shop                 72
bar                           51
american restaurant           45
chinese restaurant            40
meal delivery                 22
vietnamese restaurant         20
japanese restaurant           20
breakfast restaurant          19
indian restaurant             18
thai restaurant               12
middle eastern restaurant     10
asian restaurant              10
greek restaurant               9
mexican restaurant             8
meal takeaway                  7
seafood restaurant             6
vegan restaurant               6
korean restaurant              5
bakery and dessert             4
mediterranean restaurant       2
french restaurant              2
buffet restaurant              2
Name: count, dtype: int64

#### Define High-Level main_type Categories
While `cuisine_group` improves interpretability, the number of categories is still relatively large for high-level exploratory analysis. Therefore, a second level of grouping was introduced to classify restaurants into broader `main_type` segments based on cuisine style or business model.

The following high-level categories were defined:
- Western cuisine (e.g. American, Italian, French, Greek)
- Asian cuisine (e.g. Japanese, Chinese, Vietnamese, Indian)
- Beverage and bar (e.g. bars, cafés, bakeries)
- Fast or casual dining (e.g. fast food, buffet, breakfast restaurants)
- Delivery or takeaway
- Other cuisine (categories with relatively small sample sizes)

Restaurants that could not be clearly assigned to any of these groups were labeled as generic restaurant.

This hierarchical grouping strategy balances analytical clarity and flexibility:
- primaryType preserves granular information
- cuisine_group enables cuisine-level comparison
- main_type supports robust, high-level exploratory analysis

In [20]:
western = ['american restaurant', 'italian restaurant', 'french restaurant','greek restaurant', 'mediterranean restaurant']
asian = ['japanese restaurant', 'chinese restaurant', 'vietnamese restaurant', 'indian restaurant', 'thai restaurant',
         'korean restaurant', 'asian restaurant']
beverage_and_bar = ['bar', 'beverage shop','bakery and dessert']
fast_casual = ['fast food restaurant', 'breakfast restaurant', 'buffet restaurant']
delivery_takeaway = ['meal delivery', 'meal takeaway']
other = ['mexican restaurant', 'middle eastern restaurant', 'seafood restaurant', 'vegan restaurant']

# Define the main type of the restaurant according to the cuisine group
df.loc[df['cuisine_group'].isin(western), 'main_type'] = 'western cuisine'
df.loc[df['cuisine_group'].isin(asian), 'main_type'] = 'asian cuisine'
df.loc[df['cuisine_group'].isin(beverage_and_bar), 'main_type'] = 'beverage and bar'
df.loc[df['cuisine_group'].isin(fast_casual), 'main_type'] = 'fast or casual dining'
df.loc[df['cuisine_group'].isin(delivery_takeaway), 'main_type'] = 'delivery or takeaway'
df.loc[df['cuisine_group'].isin(other), 'main_type'] = 'other cuisine'
df['main_type'] = df['main_type'].fillna('generic restaurant')
df['main_type'].value_counts()

main_type
fast or casual dining    192
western cuisine          184
beverage and bar         127
asian cuisine            125
generic restaurant        95
other cuisine             30
delivery or takeaway      29
Name: count, dtype: int64

## 8. Extract Data from Dictionary Columns

Some columns contained dictionary-like structures. Relevant fields were extracted and expanded into separate columns, transforming semi-structured data into an analyzable tabular format.

Unnecessary raw dictionary columns were removed after extraction.

### Display Name Column

In [21]:
df['name'] = df['displayName'].apply(lambda x: x.get('text') if isinstance(x, dict) else None)
df['name'].head()

0    Ecity Flame (Best Offer on Website)
2                       JT's Bar & Grill
3                   Little Caesars Pizza
5          The Soban sushi & asian taste
6                           Panago Pizza
Name: name, dtype: object

In [22]:
df.drop(columns='displayName', inplace=True)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,paymentOptions,parkingOptions,priceRange,possible_types,cuisine_group,main_type,name
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'paidParkingLot': Fal...","{'startPrice': {'currencyCode': 'CAD', 'units'...",breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website)
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,bar,beverage and bar,JT's Bar & Grill
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,italian restaurant,western cuisine,Little Caesars Pizza
5,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste
6,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,"{'acceptsCreditCards': True, 'acceptsDebitCard...","{'freeParkingLot': True, 'freeStreetParking': ...",None,NaN,meal delivery,delivery or takeaway,Panago Pizza


### Payment Options Column

In [23]:
payment_opt = df['paymentOptions'].apply(pd.Series).fillna(False)
payment_opt.head()

,acceptsCreditCards,acceptsDebitCards,acceptsCashOnly,acceptsNfc
0,True,True,False,True
2,True,True,False,False
3,True,True,False,True
5,True,True,False,True
6,True,True,False,True


In [24]:
df['accepts_non_cash'] = payment_opt['acceptsCreditCards'] | payment_opt['acceptsDebitCards'] | payment_opt['acceptsNfc']
df.drop(columns='paymentOptions', inplace=True)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,parkingOptions,priceRange,possible_types,cuisine_group,main_type,name,accepts_non_cash
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,"{'freeParkingLot': True, 'paidParkingLot': Fal...","{'startPrice': {'currencyCode': 'CAD', 'units'...",breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,"{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,bar,beverage and bar,JT's Bar & Grill,True
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,"{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,italian restaurant,western cuisine,Little Caesars Pizza,True
5,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,"{'freeParkingLot': True, 'freeStreetParking': ...","{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True
6,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,"{'freeParkingLot': True, 'freeStreetParking': ...",None,NaN,meal delivery,delivery or takeaway,Panago Pizza,True


### Parking Options Column

In [25]:
parking_opt = df['parkingOptions'].apply(pd.Series).fillna(False)
parking_opt.head()

,freeParkingLot,paidParkingLot,freeStreetParking,paidStreetParking,valetParking,freeGarageParking,paidGarageParking
0,True,False,True,False,False,False,False
2,True,False,True,False,False,False,False
3,True,False,True,False,False,False,False
5,True,False,True,False,False,False,False
6,True,False,True,False,False,False,False


In [26]:
df['has_free_parking'] = parking_opt['freeParkingLot'] | parking_opt['freeStreetParking'] | parking_opt['freeGarageParking']
df['has_paid_parking'] = parking_opt['paidParkingLot'] | parking_opt['paidStreetParking'] | parking_opt['paidGarageParking']
df.drop(columns='parkingOptions', inplace=True)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,priceRange,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,"{'startPrice': {'currencyCode': 'CAD', 'units'...",breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True,True,False
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,"{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,bar,beverage and bar,JT's Bar & Grill,True,True,False
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,"{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,italian restaurant,western cuisine,Little Caesars Pizza,True,True,False
5,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,"{'startPrice': {'currencyCode': 'CAD', 'units'...",NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True,True,False
6,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,None,NaN,meal delivery,delivery or takeaway,Panago Pizza,True,True,False


### Price Range Column

In [27]:
price_range_df = df['priceRange'].apply(pd.Series)
for col in price_range_df:
    price_range_df[col] = price_range_df[col].apply(lambda x: x.get('units') if isinstance(x, dict) else None).astype(float)

df = pd.concat([df, price_range_df], axis=1)
df.drop(columns='priceRange', inplace=True)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True,True,False,10.0,20.0
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,NaN,bar,beverage and bar,JT's Bar & Grill,True,True,False,20.0,30.0
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,NaN,italian restaurant,western cuisine,Little Caesars Pizza,True,True,False,10.0,20.0
5,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True,True,False,20.0,30.0
6,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,NaN,meal delivery,delivery or takeaway,Panago Pizza,True,True,False,NaN,NaN


### Regular Opening Hours Column

In [28]:
opening_hrs_df = df['regularOpeningHours'].apply(lambda x: x.get('periods') if isinstance(x, dict) else None).to_frame()
opening_hrs_df.head()

,regularOpeningHours
0,"[{'open': {'day': 0, 'hour': 8, 'minute': 0}, ..."
2,"[{'open': {'day': 0, 'hour': 10, 'minute': 0},..."
3,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},..."
5,"[{'open': {'day': 0, 'hour': 12, 'minute': 0},..."
6,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},..."


The data in the regularOpeningHours column is a list of dictionaries that contained the open and close time of a period. Thus, the following functions are created in order to calculate the opening hours in each day of a week and to identify if it opens 24 hours.

In [29]:
WEEKDAY_MAP = {0: 'Sunday', 1: 'Monday', 2: 'Tuesday', 3: 'Wednesday', 4: 'Thursday', 5: 'Friday', 6: 'Saturday'}
WEEK = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
WEEKDAY = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
WEEKEND = ['Sunday', 'Saturday']

def get_weekday(time_dict):
    return WEEKDAY_MAP[time_dict['day']]

def get_timedelta(time_dict):
    return pd.Timedelta(hours=time_dict['hour'], minutes=time_dict['minute'])

# Calculate the opening hours of a day
def calculate_opening_hrs(open_period_dict):
    open_time = get_timedelta(open_period_dict['open'])
    close_time = get_timedelta(open_period_dict['close'])
    duration = close_time - open_time
    
    if duration < pd.Timedelta(0,'d'):
        duration += pd.Timedelta(days=1)

    return duration.total_seconds() / 3600

# Calculate the daily operating hours of a restaurant
def calculate_weekday_operating_hours(raw_open_periods):
    if not raw_open_periods: return None

    result = {'Sunday': 0, 'Monday': 0, 'Tuesday':0, 'Wednesday':0, 'Thursday':0, 'Friday':0, 'Saturday':0}
    for open_period_dict in raw_open_periods:
        if 'open' not in open_period_dict or 'close' not in open_period_dict:
            continue

        weekday = get_weekday(open_period_dict['open'])
        opening_hours = calculate_opening_hrs(open_period_dict)
        result[weekday] += opening_hours
    return result

In [30]:
# Check if a restaurant is open 24 hours according to the operation time
def is_open_24_hours(operation_time):
    if not isinstance(operation_time, list):
        return False

    if len(operation_time) != 1:
        return False

    entry = operation_time[0]

    return (
        'open' in entry and
        'close' not in entry and
        entry['open'].get('hour') == 0 and
        entry['open'].get('minute') == 0
    )

In [31]:
# Calculate operating_hours
daily_open_hrs_df = opening_hrs_df['regularOpeningHours'].apply(calculate_weekday_operating_hours).apply(pd.Series)
opening_hrs_df = pd.concat([opening_hrs_df, daily_open_hrs_df], axis=1)

# Check if open 24 hours
opening_hrs_df['open_24hrs'] = opening_hrs_df['regularOpeningHours'].apply(is_open_24_hours)
opening_hrs_df.loc[opening_hrs_df['open_24hrs'], WEEK] = 24
opening_hrs_df.head()

,regularOpeningHours,Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday,open_24hrs
0,"[{'open': {'day': 0, 'hour': 8, 'minute': 0}, ...",13.0,13.0,13.0,13.0,13.0,14.0,14.0,False
2,"[{'open': {'day': 0, 'hour': 10, 'minute': 0},...",14.0,13.0,15.0,15.0,13.0,16.0,16.0,False
3,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},...",11.0,11.0,11.0,11.0,12.0,13.5,12.5,False
5,"[{'open': {'day': 0, 'hour': 12, 'minute': 0},...",9.0,7.5,7.5,7.5,7.5,9.5,11.5,False
6,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},...",13.0,13.0,13.0,13.0,13.0,14.0,14.0,False


The opening hours information was extracted and transformed into meaningful features to better represent restaurant availability patterns.

In [32]:
# Weekday & Weekend Open Days
opening_hrs_df['weekday_open_days'] = (opening_hrs_df[WEEKDAY] > 0).sum(axis=1)
opening_hrs_df['weekend_open_days'] = (opening_hrs_df[WEEKEND] > 0).sum(axis=1)

# Total Weekly Opening Hours
opening_hrs_df['total_hrs'] = opening_hrs_df[WEEK].sum(axis=1)

# Avg. Daily, Weekday, Weekend Hours
opening_hrs_df['avg_daily_hrs'] = opening_hrs_df['total_hrs'] / (opening_hrs_df['weekday_open_days'] + opening_hrs_df['weekend_open_days'])
opening_hrs_df['avg_weekday_hrs'] = opening_hrs_df[WEEKDAY].sum(axis=1) / opening_hrs_df['weekday_open_days']
opening_hrs_df['avg_weekend_hrs'] = opening_hrs_df[WEEKEND].sum(axis=1) / opening_hrs_df['weekend_open_days']

opening_hrs_df.head()

,regularOpeningHours,Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs
0,"[{'open': {'day': 0, 'hour': 8, 'minute': 0}, ...",13.0,13.0,13.0,13.0,13.0,14.0,14.0,False,5,2,93.0,13.285714,13.2,13.50
2,"[{'open': {'day': 0, 'hour': 10, 'minute': 0},...",14.0,13.0,15.0,15.0,13.0,16.0,16.0,False,5,2,102.0,14.571429,14.4,15.00
3,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},...",11.0,11.0,11.0,11.0,12.0,13.5,12.5,False,5,2,82.0,11.714286,11.7,11.75
5,"[{'open': {'day': 0, 'hour': 12, 'minute': 0},...",9.0,7.5,7.5,7.5,7.5,9.5,11.5,False,5,2,60.0,8.571429,7.9,10.25
6,"[{'open': {'day': 0, 'hour': 11, 'minute': 0},...",13.0,13.0,13.0,13.0,13.0,14.0,14.0,False,5,2,93.0,13.285714,13.2,13.50


In [33]:
# Add to the main df
opening_hrs_df = opening_hrs_df[['open_24hrs', 'weekday_open_days','weekend_open_days', 'total_hrs', 'avg_daily_hrs', 'avg_weekday_hrs','avg_weekend_hrs']]
df = pd.concat([df, opening_hrs_df], axis=1)
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True,True,False,10.0,20.0,False,5,2,93.0,13.285714,13.2,13.50
2,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,NaN,bar,beverage and bar,JT's Bar & Grill,True,True,False,20.0,30.0,False,5,2,102.0,14.571429,14.4,15.00
3,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,NaN,italian restaurant,western cuisine,Little Caesars Pizza,True,True,False,10.0,20.0,False,5,2,82.0,11.714286,11.7,11.75
5,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True,True,False,20.0,30.0,False,5,2,60.0,8.571429,7.9,10.25
6,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,NaN,meal delivery,delivery or takeaway,Panago Pizza,True,True,False,NaN,NaN,False,5,2,93.0,13.285714,13.2,13.50


## 9. Extract FSA Code from Formatted Address

The Forward Sortation Area (FSA) code was extracted from the formatted address field and then find the area with the postal code list. This enables geographic grouping and area-level analysis of restaurant locations.

The postal code list using here was data extracted from [01_get_postal_codes.ipynb](01_get_postal_codes.ipynb).

In [34]:
df['fsa_code'] = df['formattedAddress'].str.extract(r'([A-Z]\d[A-Z])\s?\d[A-Z]\d')
df['fsa_code'].head()

0    T6E
2    T6K
3    T8N
5    T6B
6    T6L
Name: fsa_code, dtype: object

In [35]:
fsa_df = pd.read_csv('data/ab_postal_codes_list.csv')
fsa_df.head()

,fsa,city,area
0,T1A,Medicine Hat,Central
1,T2A,Calgary,Penbrooke Meadows / Marlborough
2,T3A,Calgary,Dalhousie / Edgemont / Hamptons / Hidden Valley
3,T4A,Airdrie,East
4,T5A,Edmonton,WestClareview / EastLondonderry


In [36]:
df = pd.merge(df, fsa_df, how='left', left_on='fsa_code', right_on='fsa').drop(columns='fsa')
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs,fsa_code,city,area
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True,True,False,10.0,20.0,False,5,2,93.0,13.285714,13.2,13.50,T6E,Edmonton,SouthBonnie Doon / EastUniversity
1,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,NaN,bar,beverage and bar,JT's Bar & Grill,True,True,False,20.0,30.0,False,5,2,102.0,14.571429,14.4,15.00,T6K,Edmonton,WestMill Woods
2,ChIJD_onNiskoFMRz9s6W7Si_Qw,"[pizza_restaurant, fast_food_restaurant, meal_...","2 Hebert Rd, St. Albert, AB T8N 6T7, Canada",3.3,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,84.0,pizza restaurant,False,True,False,NaN,italian restaurant,western cuisine,Little Caesars Pizza,True,True,False,10.0,20.0,False,5,2,82.0,11.714286,11.7,11.75,T8N,St. Albert,NaN
3,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True,True,False,20.0,30.0,False,5,2,60.0,8.571429,7.9,10.25,T6B,Edmonton,SECapilano / WestSoutheast Industrial / East...
4,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,NaN,meal delivery,delivery or takeaway,Panago Pizza,True,True,False,NaN,NaN,False,5,2,93.0,13.285714,13.2,13.50,T6L,Edmonton,EastMill Woods


As we want to analyze the restaurants in Edmonton only, the restaurants outside Edmonton would be filtered out.

In [37]:
df = df[df['city']=='Edmonton']
df.head()

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs,fsa_code,city,area
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,"[breakfast_restaurant, restaurant, point_of_in...","5706 75 Street NW, Edmonton, AB T6E 5X6, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",inexpensive,749.0,breakfast restaurant,True,True,True,breakfast restaurant,breakfast restaurant,fast or casual dining,Ecity Flame (Best Offer on Website),True,True,False,10.0,20.0,False,5,2,93.0,13.285714,13.2,13.50,T6E,Edmonton,SouthBonnie Doon / EastUniversity
1,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,"[bar_and_grill, bar, restaurant, point_of_inte...","1107 Knottwood Rd E Northwest, Edmonton, AB T6...",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,884.0,bar and grill,False,False,True,NaN,bar,beverage and bar,JT's Bar & Grill,True,True,False,20.0,30.0,False,5,2,102.0,14.571429,14.4,15.00,T6K,Edmonton,WestMill Woods
3,ChIJ44uK6H8YoFMR7-As8n-sr98,"[sushi_restaurant, korean_restaurant, japanese...","7450 Whyte Ave NW, Edmonton, AB T6B 0G2, Canada",4.6,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,732.0,sushi restaurant,False,True,True,NaN,japanese restaurant,asian cuisine,The Soban sushi & asian taste,True,True,False,20.0,30.0,False,5,2,60.0,8.571429,7.9,10.25,T6B,Edmonton,SECapilano / WestSoutheast Industrial / East...
4,ChIJMf4TBj4ZoFMRTrRLBranVmU,"[pizza_restaurant, meal_delivery, food_deliver...","6132 28 Ave NW, Edmonton, AB T6L 6N5, Canada",4.0,"{'openNow': True, 'periods': [{'open': {'day':...",moderate,95.0,meal delivery,False,False,False,NaN,meal delivery,delivery or takeaway,Panago Pizza,True,True,False,NaN,NaN,False,5,2,93.0,13.285714,13.2,13.50,T6L,Edmonton,EastMill Woods
5,ChIJWzGu-MMeoFMRzyGYhWapt5w,"[japanese_restaurant, sushi_restaurant, bar, r...","1903 98 St NW, Edmonton, AB T6N 1N5, Canada",4.3,"{'openNow': True, 'periods': [{'open': {'day':...",expensive,1568.0,japanese restaurant,False,True,True,NaN,japanese restaurant,asian cuisine,Mikado Southside,True,True,False,NaN,NaN,False,5,2,70.0,10.000000,10.0,10.00,T6N,Edmonton,South Industrial


## 10. Handle Missing Values
Missing values were handled based on the meaning and analytical importance of each column.
Decisions were made to either fill, retain, or remove missing values to maintain data integrity.

In [38]:
df.isna().sum().sort_values(ascending=False)[df.isna().sum() > 0]

possible_types         670
endPrice               149
startPrice             146
priceLevel             116
avg_weekend_hrs         42
regularOpeningHours     20
avg_weekday_hrs         20
avg_daily_hrs           20
userRatingCount          4
rating                   4
dtype: int64

### `regularOpeningHours`, `userRatingCount`, `rating` Columns
Since the subsequent analysis focuses on factors associated with restaurant success, including ratings and operating hours, rows with missing values in these key fields are removed. Retaining incomplete records could introduce noise or bias in later
analysis.

In [39]:
df.dropna(subset=['regularOpeningHours', 'userRatingCount', 'rating'], inplace=True)

In [40]:
df.isna().sum().sort_values(ascending=False)[df.isna().sum() > 0]

possible_types     653
endPrice           136
startPrice         133
priceLevel         105
avg_weekend_hrs     22
dtype: int64

### `avg_weekend_hrs` Column
Values in the avg_weekend_hrs are missing because the restaurants did not open on weekend. Thus, simply fill 0 in the missing values.

In [41]:
df[df['avg_weekend_hrs'].isna()]

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs,fsa_code,city,area
17,ChIJYUnYxjUioFMRdctU3nuy4YM,"[cafe, fast_food_restaurant, coffee_shop, food...","MacEwan University, 10800 105 Ave NW Second Fl...",3.0,"{'openNow': False, 'periods': [{'open': {'day'...",inexpensive,254.0,cafe,True,False,False,NaN,beverage shop,beverage and bar,Tim Hortons,True,True,False,1.0,10.0,False,5,0,47.5,9.5,9.5,NaN,T5J,Edmonton,NorthDowntown
75,ChIJz0lt7QIioFMR5j9VBIhoydc,"[coffee_shop, fast_food_restaurant, food_store...","Central Academic Building, 90 114 St NW, Edmon...",2.4,"{'openNow': False, 'periods': [{'open': {'day'...",inexpensive,97.0,coffee shop,True,False,False,NaN,beverage shop,beverage and bar,Tim Hortons,True,True,False,1.0,10.0,False,5,0,37.5,7.5,7.5,NaN,T6G,Edmonton,WestUniversity / Strathcona Place
85,ChIJe4h6y0ofoFMRKhD61wUuAJE,"[fast_food_restaurant, food_store, restaurant,...","5041 Gateway Blvd NW, Edmonton, AB T6H 4R7, Ca...",4.2,"{'openNow': False, 'periods': [{'open': {'day'...",None,5.0,fast food restaurant,False,False,False,NaN,fast food restaurant,fast or casual dining,Chiro Foods Limited,False,True,False,NaN,NaN,False,5,0,40.0,8.0,8.0,NaN,T6H,Edmonton,Southgate / NorthRiverbend
135,ChIJK9qG-mQZoFMRsUdxUlMhwsE,"[breakfast_restaurant, restaurant, point_of_in...","3615 24 Ave NW, Edmonton, AB T6L 4V5, Canada",5.0,"{'openNow': True, 'periods': [{'open': {'day':...",None,4.0,breakfast restaurant,False,False,False,NaN,breakfast restaurant,fast or casual dining,Doaba indian tiffin service,False,False,False,NaN,NaN,False,4,0,44.0,11.0,11.0,NaN,T6L,Edmonton,EastMill Woods
168,ChIJBbyIml4hoFMRPmFmnssg-AE,"[restaurant, point_of_interest, food, establis...","14413 115 Ave NW, Edmonton, AB T5M 3B9, Canada",4.7,"{'openNow': False, 'periods': [{'open': {'day'...",inexpensive,208.0,restaurant,False,False,True,NaN,restaurant,generic restaurant,Lunchbox Diner,True,True,False,10.0,20.0,False,5,0,25.0,5.0,5.0,NaN,T5M,Edmonton,SouthWestmount / Groat Estate / EastNorthwest...
248,ChIJ6ZM3OLYYoFMRo3xdrK4JURc,"[sandwich_shop, restaurant, food, point_of_int...","9704 54 Ave, Edmonton, AB T6E 0A9, Canada",4.8,"{'openNow': False, 'periods': [{'open': {'day'...",inexpensive,135.0,sandwich shop,False,False,False,NaN,fast food restaurant,fast or casual dining,Cafe On 54th,True,True,False,10.0,20.0,False,5,0,35.0,7.0,7.0,NaN,T6E,Edmonton,SouthBonnie Doon / EastUniversity
281,ChIJpzrMgv4eoFMR9qD1-5qc-oI,"[meal_delivery, food_delivery, food_store, res...","845 Saddleback Rd NW, Edmonton, AB T6J 4W4, Ca...",5.0,"{'openNow': False, 'periods': [{'open': {'day'...",None,2.0,meal delivery,False,False,False,NaN,meal delivery,delivery or takeaway,The Second Chef,True,True,False,NaN,NaN,False,5,0,40.0,8.0,8.0,NaN,T6J,Edmonton,Kaskitayo / Aspen Gardens
282,ChIJMYBT1vYhoFMR-pAmSeLkAGQ,"[coffee_shop, donut_shop, breakfast_restaurant...",1045 Katz Group - Rexall Centre Of Pharmacy An...,4.0,"{'openNow': False, 'periods': [{'open': {'day'...",inexpensive,171.0,coffee shop,True,False,False,NaN,beverage shop,beverage and bar,Tim Hortons,True,True,False,1.0,10.0,False,5,0,40.0,8.0,8.0,NaN,T6G,Edmonton,WestUniversity / Strathcona Place
342,ChIJl1MS2R8ioFMRMgR0QBcc3lA,"[restaurant, point_of_interest, food, establis...","University of Alberta, Education Centre, 87 Av...",4.4,"{'openNow': False, 'periods': [{'open': {'day'...",None,12.0,restaurant,False,False,True,NaN,restaurant,generic restaurant,EuroMarket,True,False,False,NaN,NaN,False,5,0,43.5,8.7,8.7,NaN,T6G,Edmonton,WestUniversity / Strathcona Place
370,ChIJZ-z3aiAioFMR1hvyKZsBtiI,"[restaurant, asian_restaurant, korean_restaura...","University of Alberta, 8900 114 St NW Basement...",4.8,"{'openNow': True, 'perio

In [42]:
df.loc[df['weekend_open_days'] == 0 & df['avg_weekend_hrs'].isna(), 'avg_weekend_hrs'] = 0

### `priceLevel` Column
Filling the missing price level with the price range (i.e. `startPrice` and `endPrice`) of the restaurants.
Using the most frequent start price and end price in each price level as reference. <br>
For data without price level and price range, the records would be removed as they could not support meaningful analysis.

In [43]:
df.groupby('priceLevel')[['startPrice', 'endPrice']].agg(pd.Series.mode)

,startPrice,endPrice
priceLevel,,
expensive,40.0,50.0
inexpensive,10.0,20.0
moderate,20.0,30.0
very expensive,100.0,[]


In [44]:
df.loc[df['endPrice']<= 20, 'priceLevel'] = df.loc[df['endPrice']<= 20].fillna('inexpensive')
df.loc[df['endPrice']<= 40, 'priceLevel'] = df.loc[df['endPrice']<= 40].fillna('moderate')

df[df['priceLevel'].isna() & (df['startPrice'].notna() | df['endPrice'].notna())]

,id,types,formattedAddress,rating,regularOpeningHours,priceLevel,userRatingCount,primaryType,menuForChildren,goodForChildren,goodForGroups,possible_types,cuisine_group,main_type,name,accepts_non_cash,has_free_parking,has_paid_parking,startPrice,endPrice,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs,fsa_code,city,area


In [45]:
df.dropna(subset='priceLevel', inplace=True)

## 11. Final Column Cleanup and Reordering
After completing all transformations, remaining unnecessary columns were removed.
The columns were then rearranged to improve readability and support downstream analysis.

In [46]:
df.columns

Index(['id', 'types', 'formattedAddress', 'rating', 'regularOpeningHours',
       'priceLevel', 'userRatingCount', 'primaryType', 'menuForChildren',
       'goodForChildren', 'goodForGroups', 'possible_types', 'cuisine_group',
       'main_type', 'name', 'accepts_non_cash', 'has_free_parking',
       'has_paid_parking', 'startPrice', 'endPrice', 'open_24hrs',
       'weekday_open_days', 'weekend_open_days', 'total_hrs', 'avg_daily_hrs',
       'avg_weekday_hrs', 'avg_weekend_hrs', 'fsa_code', 'city', 'area'],
      dtype='object')

In [47]:
df = df[['id', 'name', 'fsa_code', 'area', 'main_type', 'cuisine_group', 'priceLevel', 'rating','userRatingCount', 'menuForChildren',
         'goodForChildren','goodForGroups', 'accepts_non_cash', 'has_free_parking', 'has_paid_parking', 'open_24hrs', 'weekday_open_days',
         'weekend_open_days','total_hrs', 'avg_daily_hrs', 'avg_weekday_hrs', 'avg_weekend_hrs']]

df.rename(columns={'priceLevel': 'price_level', 'userRatingCount': 'user_rating_count', 'menuForChildren': 'menu_for_children',
                   'goodForChildren': 'good_for_children', 'goodForGroups': 'good_for_groups'}, inplace=True)

df.head()

,id,name,fsa_code,area,main_type,cuisine_group,price_level,rating,user_rating_count,menu_for_children,good_for_children,good_for_groups,accepts_non_cash,has_free_parking,has_paid_parking,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs
0,ChIJQWNpDe0YoFMRto7xgtIAHEk,Ecity Flame (Best Offer on Website),T6E,SouthBonnie Doon / EastUniversity,fast or casual dining,breakfast restaurant,inexpensive,4.0,749.0,True,True,True,True,True,False,False,5,2,93.0,13.285714,13.2,13.50
1,ChIJt4Bu_ksZoFMRR0pb8OvRXoQ,JT's Bar & Grill,T6K,WestMill Woods,beverage and bar,bar,moderate,4.3,884.0,False,False,True,True,True,False,False,5,2,102.0,14.571429,14.4,15.00
3,ChIJ44uK6H8YoFMR7-As8n-sr98,The Soban sushi & asian taste,T6B,SECapilano / WestSoutheast Industrial / East...,asian cuisine,japanese restaurant,moderate,4.6,732.0,False,True,True,True,True,False,False,5,2,60.0,8.571429,7.9,10.25
4,ChIJMf4TBj4ZoFMRTrRLBranVmU,Panago Pizza,T6L,EastMill Woods,delivery or takeaway,meal delivery,moderate,4.0,95.0,False,False,False,True,True,False,False,5,2,93.0,13.285714,13.2,13.50
5,ChIJWzGu-MMeoFMRzyGYhWapt5w,Mikado Southside,T6N,South Industrial,asian cuisine,japanese restaurant,expensive,4.3,1568.0,False,True,True,True,True,False,False,5,2,70.0,10.000000,10.0,10.00


## 12. Data Cleaning Summary

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 781
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 712 non-null    object 
 1   name               712 non-null    object 
 2   fsa_code           712 non-null    object 
 3   area               712 non-null    object 
 4   main_type          712 non-null    object 
 5   cuisine_group      712 non-null    object 
 6   price_level        712 non-null    object 
 7   rating             712 non-null    float64
 8   user_rating_count  712 non-null    float64
 9   menu_for_children  712 non-null    bool   
 10  good_for_children  712 non-null    bool   
 11  good_for_groups    712 non-null    bool   
 12  accepts_non_cash   712 non-null    bool   
 13  has_free_parking   712 non-null    bool   
 14  has_paid_parking   712 non-null    bool   
 15  open_24hrs         712 non-null    bool   
 16  weekday_open_days  712 non-null

In [49]:
df.describe(include='all')

,id,name,fsa_code,area,main_type,cuisine_group,price_level,rating,user_rating_count,menu_for_children,good_for_children,good_for_groups,accepts_non_cash,has_free_parking,has_paid_parking,open_24hrs,weekday_open_days,weekend_open_days,total_hrs,avg_daily_hrs,avg_weekday_hrs,avg_weekend_hrs
count,712,712,712,712,712,712,712,712.000000,712.000000,712,712,712,712,712,712,712,712.000000,712.000000,712.000000,712.000000,712.000000,712.000000
unique,712,393,37,37,7,25,4,NaN,NaN,2,2,2,2,2,2,2,NaN,NaN,NaN,NaN,NaN,NaN
top,ChIJQWNpDe0YoFMRto7xgtIAHEk,Tim Hortons,T5T,West Jasper Place / West Edmonton Mall,fast or casual dining,fast food restaurant,inexpensive,NaN,NaN,True,True,True,True,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,57,48,48,182,163,356,NaN,NaN,405,519,509,710,678,673,652,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.022051,874.824438,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.905899,1.905899,86.676030,12.612028,12.706461,12.175152
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.475480,1104.347648,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.328435,0.361072,33.364905,4.558515,4.547809,5.047230
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.200000,5.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.000000,0.000000,20.000000,3.428571,3.500000,0.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.800000,271.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,2.000000,66.500000,10.000000,10.000000,9.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.100000,548.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,2.000000,83.000000,11.910714,12.000000,11.625000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.400000,1066.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,2.000000,99.000000,14.142857,14.200000,14.000000


After cleaning, the dataset contains:
- 712 restaurants
- 22 columns

The dataset is transformed from the raw Google Maps export into a structured and analysis-ready format. Key steps include:
- Convert data types
- Extracting and standardizing nested attributes
- Engineering features related to weekday and weekend operations
- Handling missing values in critical fields

## 13. Export Cleaned Data
After the data cleaning process, the dataset is well-structured, consistent, and ready for exploratory data analysis and modeling.

In [50]:
df.to_csv('data/cleaned_restaurants.csv', index=False)